test/cpp/microbenchmarks/callback_test_service.h

```diff
@@ -0,0 +1,130 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef TEST_CPP_MICROBENCHMARKS_CALLBACK_TEST_SERVICE_H
+#define TEST_CPP_MICROBENCHMARKS_CALLBACK_TEST_SERVICE_H
+
+#include <benchmark/benchmark.h>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/cpp/util/string_ref_helper.h"
+
+namespace grpc {
+namespace testing {
+
+const char* const kServerFinishAfterNReads = "server_finish_after_n_reads";
+const char* const kServerMessageSize = "server_message_size";
+
+class CallbackStreamingTestService
+    : public EchoTestService::ExperimentalCallbackService {
+ public:
+  CallbackStreamingTestService() {}
+  void Echo(ServerContext* context, const EchoRequest* request,
+            EchoResponse* response,
+            experimental::ServerCallbackRpcController* controller) override;
+
+  experimental::ServerBidiReactor<EchoRequest, EchoResponse>* BidiStream()
+      override;
+};
+
+class BidiClient
+    : public grpc::experimental::ClientBidiReactor<EchoRequest, EchoResponse> {
+ public:
+  BidiClient(benchmark::State* state, EchoTestService::Stub* stub,
+             EchoRequest* request, EchoResponse* response, std::mutex* mu,
+             std::condition_variable* cv, bool* done)
+      : state_{state},
+        stub_{stub},
+        request_{request},
+        response_{response},
+        mu_{mu},
+        cv_{cv},
+        done_(done) {
+    msgs_size_ = state->range(0);
+    msgs_to_send_ = state->range(1);
+  }
+
+  void OnReadDone(bool ok) override {
+    if (!ok) {
+      return;
+    }
+    if (writes_complete_ < msgs_to_send_) {
+      MaybeWrite();
+    }
+  }
+
+  void OnWriteDone(bool ok) override {
+    if (!ok) {
+      return;
+    }
+    writes_complete_++;
+    StartRead(response_);
+  }
+
+  void OnDone(const Status& s) override {
+    GPR_ASSERT(s.ok());
+    if (state_->KeepRunning()) {
+      BidiClient* test =
+          new BidiClient(state_, stub_, request_, response_, mu_, cv_, done_);
+      test->StartNewRpc();
+    } else {
+      std::unique_lock<std::mutex> l(*mu_);
+      *done_ = true;
+      cv_->notify_one();
+    }
+    delete cli_ctx_;
```

Comment:
This should likely move before StartNewRPC since otherwise StartNewRPC will leak the existing one and then this one will get deleted instead